main.go

```diff
@@ -156,78 +155,24 @@ func metricsServer() {
-
-func (l DeploymentLister) List() ([]v1beta1.Deployment, error) {
-	return l()
-}
-
-type PodLister func() ([]v1.Pod, error)
-
-func (l PodLister) List() ([]v1.Pod, error) {
-	return l()
-}
-
-type NodeLister func() (v1.NodeList, error)
-
-func (l NodeLister) List() (v1.NodeList, error) {
-	return l()
-}
-
 // initializeMetrics creates a new controller from the given config.
 func initializeMetrics(kubeClient clientset.Interface) {
-	dplStore, dplController := cache.NewNamespaceKeyedIndexerAndReflector(
-		&cache.ListWatch{
-			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
-				return kubeClient.Extensions().Deployments(api.NamespaceAll).List(options)
-			},
-			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
-				return kubeClient.Extensions().Deployments(api.NamespaceAll).Watch(options)
-			},
-		}, &v1beta1.Deployment{}, resyncPeriod)
-
-	podStore, podController := cache.NewNamespaceKeyedIndexerAndReflector(
-		cache.NewListWatchFromClient(
-			kubeClient.Core().GetRESTClient(),
-			"pods",
-			api.NamespaceAll,
-			nil,
-		), &v1.Pod{}, resyncPeriod)
-
-	nodeStore, nodeController := cache.NewNamespaceKeyedIndexerAndReflector(
-		cache.NewListWatchFromClient(
-			kubeClient.Core().GetRESTClient(),
-			"nodes",
-			api.NamespaceAll,
-			nil,
-		), &v1.Node{}, resyncPeriod)
-
-	go dplController.Run()
-	go podController.Run()
-	go nodeController.Run()
-
-	dplLister := DeploymentLister(func() (deployments []v1beta1.Deployment, err error) {
-		for _, c := range dplStore.List() {
-			deployments = append(deployments, *(c.(*v1beta1.Deployment)))
-		}
-		return deployments, nil
-	})
+	cclient := kubeClient.Core().GetRESTClient()
+	eclient := kubeClient.Extensions().GetRESTClient()
 
-	podLister := PodLister(func() (pods []v1.Pod, err error) {
-		for _, m := range podStore.List() {
-			pods = append(pods, *m.(*v1.Pod))
-		}
-		return pods, nil
-	})
+	dlw := cache.NewListWatchFromClient(eclient, "deployments", api.NamespaceAll, nil)
+	plw := cache.NewListWatchFromClient(cclient, "pods", api.NamespaceAll, nil)
+	nlw := cache.NewListWatchFromClient(cclient, "nodes", api.NamespaceAll, nil)
 
-	nodeLister := NodeLister(func() (machines v1.NodeList, err error) {
-		for _, m := range nodeStore.List() {
-			machines.Items = append(machines.Items, *(m.(*v1.Node)))
-		}
-		return machines, nil
-	})
+	dinf := cache.NewSharedInformer(dlw, &v1beta1.Deployment{}, resyncPeriod)
+	pinf := cache.NewSharedInformer(plw, &v1.Pod{}, resyncPeriod)
+	ninf := cache.NewSharedInformer(nlw, &v1.Node{}, resyncPeriod)
+
+	prometheus.MustRegister(&deploymentCollector{store: dinf.GetStore()})
+	prometheus.MustRegister(&podCollector{store: pinf.GetStore()})
+	prometheus.MustRegister(&nodeCollector{store: ninf.GetStore()})
 
-	prometheus.MustRegister(&deploymentCollector{store: dplLister})
-	prometheus.MustRegister(&podCollector{store: podLister})
-	prometheus.MustRegister(&nodeCollector{store: nodeLister})
+	go dinf.Run(context.TODO().Done())
```

Comment:
We can just use background here as this is basically top-level. From my understanding, TODO is intended for when you want to connect a context chain later on but can't yet because the call chain doesn't fully support it yet.